[clang][analyzer] Handle CXXParenInitListExpr alongside InitListExpr

fangyi-zhou opened the pull request:
As reported in #135665, C++20 parenthesis initializer list expressions are not handled correctly and were causing crashes. This commit attempts to fix the issue by handing parenthesis initializer lists along side existing initializer lists.

zwuis commented on clang/lib/StaticAnalyzer/Checkers/DynamicTypePropagation.cpp:
```diff
@@ -379,10 +379,12 @@ void DynamicTypePropagation::checkPostCall(const CallEvent &Call,
+        // available in this callback, ideally as part of CallEvent.
+        const Stmt *Parent =
+            LCtx->getParentMap().getParent(Ctor->getOriginExpr());
+        if (isa_and_nonnull<InitListExpr>(Parent) ||
+            isa_and_nonnull<CXXParenListInitExpr>(Parent))
```
```cpp
isa_and_nonnull<Expr1, Expr2>(...)
```

Ditto below.

zwuis reviewed (commented):
Thank you for your patch! Please add a release note in clang/docs/ReleaseNotes.rst .

zwuis commented on clang/docs/ReleaseNotes.rst:
There is a section named "Static Analyzer" in this file.

fangyi-zhou replied:
Sorry I didn't notice, will fix.

fangyi-zhou committed:
[clang][analyzer] Handle CXXParenInitListExpr alongside InitListExpr

As reported in #135665, C++20 parenthesis initializer list expressions
are not handled correctly and were causing crashes. This commit attempts
to fix the issue by handing parenthesis initializer lists along side
existing initializer lists.

zwuis commented:
@Xazax-hun @steakhal Please take a look.

steakhal commented on clang/test/Analysis/PR135665.cpp:
```diff
@@ -0,0 +1,19 @@
+
+int main()
+{
+  const auto l = overload([](const int* i) {});
```
```suggestion
  const auto l = overload([](const int* i) {}); // no-crash
```

steakhal reviewed (approved):
LGTM, thanks. FYI we usually ping once a week.

steakhal committed:
Add no-crash in test

zwuis commented:
> FYI we usually ping once a week.

Oh, I see. IIUC `@ llvm/pr-subscribers-clang-static-analyzer-1` is also a ping, which I didn't know.

steakhal commented:
Ah, I see now your perspective.
Yes, any PRs tagged with the CSA tag will hit our inboxes.
